swap/swap.go

```diff
@@ -436,51 +470,58 @@ func (s *Swap) getContractOwner(ctx context.Context, address common.Address) (co
+func (s *Swap) StartChequebook(chequebookAddrFlag common.Address) error {
+	chequebookToUse, err := s.loadChequebook()
+	// error reading from disk
+	if err != nil && err != state.ErrNotFound {
+		return fmt.Errorf("Error reading previously used chequebook: %s", err)
+	}
+	// nothing written to state disk before, no flag provided: deploying new chequebook
+	if err == state.ErrNotFound && chequebookAddrFlag == (common.Address{}) {
+		chequebook, err := s.Deploy(context.TODO())
+		if err != nil {
+			return fmt.Errorf("Error deploying chequebook: %v", err)
 		}
-		log.Info("New SWAP contract deployed", "contract info", s.DeploySuccess())
+		chequebookToUse = chequebook
 	}
-	return nil
-}
 
-// BindToContractAt binds an instance of an already existing chequebook contract at address and sets chequebookAddr
-func (s *Swap) BindToContractAt(address common.Address) (err error) {
+	// read from state, but provided flag is not the same
+	if err == nil && (chequebookAddrFlag != common.Address{} && chequebookAddrFlag != chequebookToUse) {
+		return fmt.Errorf("Attempting to connect to provided chequebook, but different chequebook used before")
+	}
+	if chequebookAddrFlag != (common.Address{}) {
+		chequebookToUse = chequebookAddrFlag
+	}
+	auditLog.Info("Using the chequebook", "chequebookAddr", chequebookToUse)
+	return s.bindToContractAt(chequebookToUse)
+}
 
+// BindToContractAt binds to an instance of an already existing chequebook contract at address
+func (s *Swap) bindToContractAt(address common.Address) (err error) {
+	// validate whether address is a chequebook
 	if err := contract.ValidateCode(context.Background(), s.backend, address); err != nil {
-		return fmt.Errorf("contract validation for %v failed: %v", address, err)
+		return fmt.Errorf("contract validation for %v failed: %v", address.Hex(), err)
 	}
+	// get the instance and save it on swap.contract
 	s.contract, err = contract.InstanceAt(address, s.backend)
 	if err != nil {
 		return err
 	}
-	return nil
+	auditLog.Info("Using the chequebook", "chequebookAddr", address, "networkID", s.backendNetworkID)
+	// saving chequebook
+	return s.saveChequebook(address)
 }
 
 // Deploy deploys the Swap contract and sets the contract address
-func (s *Swap) Deploy(ctx context.Context) error {
+func (s *Swap) Deploy(ctx context.Context) (common.Address, error) {
 	opts := bind.NewKeyedTransactor(s.owner.privateKey)
 	// initial topup value
 	opts.Value = big.NewInt(int64(s.params.InitialDepositAmount))
 	opts.Context = ctx
 
-	auditLog.Info("deploying new swap", "owner", opts.From.Hex())
-	address, err := s.deployLoop(opts, s.owner.address, defaultHarddepositTimeoutDuration)
-	if err != nil {
-		auditLog.Error("unable to deploy swap", "error", err)
-		return err
-	}
-	auditLog.Info("swap deployed", "address", address.Hex(), "owner", opts.From.Hex())
```

Comment:
I would like to know as well

Reply:
The log, when we deploy a chequebook will be:

`auditLog.Info("deploying new swap", "owner", opts.From.Hex(), "deposit", opts.Value)` (line 538)
`auditLog.Info("Using the chequebook", "chequebookAddr", chequebookToUse)` (line 512)

Given that we have these two logs, I thought it would be unnecessary to also have the log 

`auditLog.Info("swap deployed", "address", address.Hex(), "owner", opts.From.Hex())`

If you don't agree, I will change it

Reply:
I think that "using the chequebook" isn't very helpful (is the chequebook address in hex?)

First "deploying new..." and then "swap deployed"  makes more sense to me. It clearly shows the user that the deployment is being attempted and then that it succeeded.

Maybe we could move "using the chequebook" to a debug log.

Reply:
Updated. 
We now have a log for deploying,
`auditLog.Info("Deploying new swap", "owner", opts.From.Hex(), "deposit", opts.Value)`
log for deployed 
`auditLog.Info("Deployed chequebook", "contract address", contract.ContractParams().ContractAddress.Hex(), "deposit", toDeposit, "owner", owner.address)`
amd a seperate log for if we are re-using a previously deployed chequebook
`auditLog.Info("bound to chequebook", "chequebookAddr", address)`